solver/src/interactions/block_coinbase.rs

```diff
@@ -0,0 +1,25 @@
+use crate::{encoding::EncodedInteraction, settlement::Interaction};
+use hex_literal::hex;
+use primitive_types::{H160, U256};
+
+// vk: A simple contract I made with verified code on etherscan:
+// https://etherscan.io/address/0x5c2cD95CF750B8f8A4881d96F04bf571A07042B1
+// Gas use for a full transaction when amount is 0 is 23849 and nonzero 30549.
+const MAINNET_ADDRESS: H160 = H160(hex!("5c2cd95cf750b8f8a4881d96f04bf571a07042b1"));
```

Comment:
This contract requires the settlement contract to hold Eth right? So we fund the Settlement contract with gas money. This is nice as we already have ways to manage the settlement contract funds.

We can also consider sending 1wei to the tipping contract as that might decrease total gas costs (@fedgiac might know better than I do).

Reply:
I think changing the account balances to a nonzero values only applies to refunds, and contract ETH balances don't matter. In any case, I'm not sure how to compute the gas cost of the call opcode as computed in the whitepaper, so I was curious enough to test it and it seems that a nonzero balance doesn't change the gas cost ([contract transactions](https://rinkeby.etherscan.io/address/0x809d1c696963266ef163d3ea10d146050b3728f9)). Something that should make the contract more efficient without actually writing bytecode is having a `receive` function that defaults to sending the money to the coinbase, in this way we should save some calldata cost.

Something I wondered is if the coinbase could ever be a contract, in which case this call might fail if the receive logic is too expensive. In the yellow paper there is nothing explicitly saying it's not possible, however it only says that the balance is updated and not that any code would be executed, so I wonder if this is a way to deposit ETH into any contract even if it's not payable.

Reply:
Good point about coinbase being contract. I looked through the flashbots documentation again and found this https://docs.flashbots.net/flashbots-auction/searchers/advanced/coinbase-payment/#managing-payments-to-coinbaseaddress-when-it-is-a-contract :

> Miners will occasionally have a smart contract listed as their block.coinbase address. This changes the expected behavior of the making payments to block.coinbase. Specifically it costs more gas to transfer ETH to block.coinbase if it is a contract than if it is an EOA, and as such many searchers will underestimate their gas consumption and their bundles will fail for miners who use contracts instead.
To handle this edge case searchers can up their gas limit to accomodate the additional payment to miners and call block.coinbase in the following way:
> ```
> block.coinbase.call{value: _ethAmountToCoinbase}(new bytes(0));
> ```
> However, searchers should be acutely aware of the risk of reentrancy attacks, as calling coinbase in this way temporarily gives execution to a third party, and typically payments to coinbase are made after checks for profit. Moreover, searchers should be aware that supporting payments to coinbase addresses that are contracts will cause their gas consumption to go up, and as a result their bundle gas price to go down. This is a tradeoff that should be considered.

I am not sure what to do about it and how common the address being a contract is. It seems like miners should avoid making it a contract precisely because of these issues.

Reply:
> so I wonder if this is a way to deposit ETH into any contract even if it's not payable

Without testing it I am guessing that yes the miner block reward could be deposited this way. I wonder if there is a contract out there that is attackable by a miner this way.

Reply:
> so I wonder if this is a way to deposit ETH into any contract even if it's not payable

`SELFDESTRUCT` will always transfer ETH without invoking the contract code (so would work for non-payable contracts).

Reply:
This discussion reminds me of an old Smart Contract Capture the Flag problem at ETHCC (2018?) where you could fund a non-payable contract with eth by sending the eth to the contract address before deployment. This means you would have to use create2 for the contract deployment to have a deterministic address, and just send the funds there before hand. 

Anyway, not useful, just reminiscing.